infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -117,172 +135,281 @@ public PollReport poll(@Nonnull Layout layout,
 
-        // We can try to scale back the network latency time after every poll round.
-        // If there are no failures, after a few rounds our polling period converges back to
-        // initPeriodDuration.
-        period = Math.max(initPeriodDuration, (period - periodDelta));
-        tuneRoutersResponseTimeout(members, routerMap, new HashSet<>(members), period);
+        PollReport latestReport = reports.get(reports.size() - 1);
 
-        // Check all responses and collect all failures.
-        if (failuresDetected) {
-            failed = members.stream()
-                    .filter(s -> responsesMap.get(s) == null
-                            || responsesMap.get(s) != (failureThreshold - 1))
-                    .collect(ImmutableSet.toImmutableSet());
-        } else {
-            failed = ImmutableSet.of();
+        return PollReport.builder()
+                .pollEpoch(epoch)
+                .connectedNodes(ImmutableSet.copyOf(connectedNodesAggregated))
+                .failedNodes(ImmutableSet.copyOf(failedNodesAggregated))
+                .wrongEpochs(ImmutableMap.copyOf(wrongEpochsAggregated))
+                .currentLayoutSlotUnFilled(isCurrentLayoutSlotUnFilled(layout, wrongEpochsAggregated.keySet()))
+                .clusterState(latestReport.getClusterState())
+                .build();
+    }
+
+    /**
+     * Tune timeouts after each poll iteration, according to list of failed and connected nodes
+     *
+     * @param router     client router
+     * @param currReport current poll report
+     * @return updated poll interval
+     */
+    private long tunePollIterationTimeouts(Map<String, IClientRouter> router, PollReport currReport) {
+        long pollIterationStartTime = System.nanoTime();
+        long pollInterval = initialPollInterval;
+
+        Set<String> allConnectedNodes = currReport.getAllConnectedNodes();
+
+        if (!currReport.getFailedNodes().isEmpty()) {
+            // Increase the inter iteration sleep time by increasing poll interval if the
+            // period time increases
+            long iterationElapsedTime = TimeUnit.MILLISECONDS.convert(
+                    System.nanoTime() - pollIterationStartTime,
+                    TimeUnit.NANOSECONDS
+            );
+            pollInterval = Math.max(initialPollInterval, period - iterationElapsedTime);
+            period = getIncreasedPeriod();
+            tuneRoutersResponseTimeout(router, allConnectedNodes, period);
         }
+        return pollInterval;
+    }
 
+    /**
+     * We can try to scale back the network latency time after every poll round.
+     * If there are no failures, after a few rounds our polling period converges back to initPeriodDuration.
+     *
+     * @param router            router
+     * @param failedNodes       list of failed nodes
+     * @param allConnectedNodes all connected nodes
+     */
+    private void pollReportTimeoutsTuning(
```

Comment:
Please rename these 2 methods :
pollReportTimeoutsTuning and tunePollIterationTimeouts.
Please rename them to something more intuitive.

Reply:
done